RabinKarpAlgorithm/RabinKarp.cs

```diff
@@ -0,0 +1,74 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace RabinKarpAlgorithm
+{
+    public class RabinKarp
+    {
+        int mod = 113;
+        public int GetPatternCount(string baseString, string pattern)
+        {
+            int baseStringLen = baseString.Length;
+            int patternLen = pattern.Length;
+
+            char[] possibleMatchChar = new char[patternLen];
+
+            int patternHash = HashMethod(pattern);
+
+            int cnt = 0;
+            int sum = 0;
+
+            int i, j = 0;
+            int pwr = patternLen - 1;
+
+            bool match = false;
+
+            while (j < patternLen)
+            {
+                possibleMatchChar[j] = baseString[j];
+                j++;
+            }
+
+            sum = HashMethod(new string(possibleMatchChar));
+            if (sum == patternHash)
+                match = pattern.Equals(new string(possibleMatchChar));
+            if (match == true)
+                cnt++;
+
+            for (i = 1; i <= baseStringLen - patternLen; i++)
+            {
+                j = 0;
+                match = false;
+
+                while (j < patternLen - 1)
+                {
+                    possibleMatchChar[j] = possibleMatchChar[j + 1];
+                    j++;
+                }
+                possibleMatchChar[patternLen - 1] = baseString[i + patternLen - 1];
+                sum = (((sum - ((baseString[i - 1] * (int)Math.Pow(2, pwr)) % mod) + mod) % mod * 2) + baseString[i + patternLen - 1]) % mod;
+
+                if (sum == patternHash)
+                    match = pattern.Equals(new string(possibleMatchChar));
+                if (match == true)
```

Comment:
Sta ako dva stringa koja su razlicita imaju istu hash vrednost?
Hash vrednost nije jedinstvena, pogotovo kad ti je modul po kojem radis 137. To znaci da ces imati 137 jedinstvenih hash vrednosti, ako se hash vrednosti nekih stringova poklapaju, a poklapace se, sta ces onda?

Reply:
Ukoliko se hash vrednosti poklope, idu na dodatnu proveru svakog karaktera, tj Equals, ukoliko nisu isti u match-u ce biti vrednost false i nece se izvrsiti inkrementacija povratne vrednosti, tj taj string nece biti racunat kao pattern.
A sto se modula tice, njega sam stavio da bih izbegao overflow (sto nisam u potpunosti uspeo, nakon 30 karaktera algoritam vise nije funkcionalan, verovatno ima veze sa najvecom mogucom vrednoscu int-a, sad mi je palo na pamet da bi mozda unsigned int radio bolje). Kontam da bih mogao da povecam vrednost modula da bih rasirio opseg, ako mi je uopste i potreban?